File: Beecrowd/6-Paradigmas/2191-Melhor_Ordem-7/2919.java
import java.util.*;

public class Main {
    public static int lowerBound(ArrayList<Integer> arr, int target) {
        int l = 0, r = arr.size();
        while (l < r) {
            int mid = (l + r) / 2;
            if (arr.get(mid) < target) l = mid + 1;
            else r = mid;
        }
        return l;
    }

    public static int LIS(int[] nums) {
        ArrayList<Integer> lis = new ArrayList<>();
        for (int num : nums) {
            int idx = lowerBound(lis, num);
            if (idx == lis.size()) lis.add(num);
            else lis.set(idx, num);
        }
        return lis.size();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (sc.hasNextInt()) {
            int n = sc.nextInt();
            int[] nums = new int[n];
            for (int i = 0; i < n; i++) nums[i] = sc.nextInt();
            System.out.println(LIS(nums));
        }
    }
}

File: Beecrowd/6-Paradigmas/2191-Melhor_Ordem-7/2919.c
#include <stdio.h>

int lower_bound(int *arr, int size, int target) {
    int l = 0, r = size;
    while (l < r) {
        int mid = (l + r) / 2;
        if (arr[mid] < target) l = mid + 1;
        else r = mid;
    }
    return l;
}

int LIS(int *nums, int n) {
    int lis[1000001], length = 0;
    for (int i = 0; i < n; i++) {
        int pos = lower_bound(lis, length, nums[i]);
        if (pos == length) lis[length++] = nums[i];
        else lis[pos] = nums[i];
    }
    return length;
}

int main() {
    int n, nums[1000001];
    while (scanf("%d", &n) != EOF) {
        for (int i = 0; i < n; i++) scanf("%d", &nums[i]);
        printf("%d\n", LIS(nums, n));
    }
    return 0;
}
